src/sentry/workflow_engine/migration_helpers/alert_rule.py

```diff
@@ -448,30 +448,60 @@ def create_workflow(
     )
 
 
-def create_detector(
+def get_detector_field_values(
     alert_rule: AlertRule,
-    project_id: int,
     data_condition_group: DataConditionGroup,
+    project_id: int | None = None,
     user: RpcUser | None = None,
-) -> Detector:
-    return Detector.objects.create(
-        project_id=project_id,
-        enabled=True,
-        created_by_id=user.id if user else None,
-        name=alert_rule.name,
-        workflow_condition_group=data_condition_group,
-        type=MetricAlertFire.slug,
-        description=alert_rule.description,
-        owner_user_id=alert_rule.user_id,
-        owner_team=alert_rule.team,
-        config={
+) -> dict[str, Any]:
+    detector_field_values = {
+        "name": alert_rule.name,
+        "description": alert_rule.description,
+        "workflow_condition_group": data_condition_group,
+        "owner_user_id": alert_rule.user_id,
+        "owner_team": alert_rule.team,
+        "config": {
             "threshold_period": alert_rule.threshold_period,
             "sensitivity": alert_rule.sensitivity,
             "seasonality": alert_rule.seasonality,
             "comparison_delta": alert_rule.comparison_delta,
             "detection_type": alert_rule.detection_type,
         },
+    }
+    if project_id is not None:
+        # these fields are only set on create, not update
+        detector_field_values.update(
+            {
+                "project_id": project_id,
+                "enabled": True,
+                "created_by_id": user.id if user else None,
+                "type": MetricAlertFire.slug,
+            }
+        )
+    return detector_field_values
+
+
+def create_detector(
+    alert_rule: AlertRule,
+    project_id: int,
+    data_condition_group: DataConditionGroup,
+    user: RpcUser | None = None,
+) -> Detector:
+    detector_field_values = get_detector_field_values(
+        alert_rule, data_condition_group, project_id, user
     )
+    return Detector.objects.create(**detector_field_values)
```

Comment:
for future / just general learnings, it's probably better if this used the `DetectorValidator` to ensure the detector_field_values are correct and provide clean validation errors etc. (not a huge deal to worry about rn though)